packages/internal/src/types.d.ts

```diff
@@ -18,3 +18,5 @@ export declare class SyncCallback<
 
   public isSync: true;
 }
+
+export type ExoObj<T> = Awaited<ReturnType<Awaited<ReturnType<T>>>>;
```

Comment:
It is not obvious to me why this pattern repeatedly appears, or at least slight simpler ones subsumed by this one. Please add a doccomment explaining what this type means and when it should be used.

Is that reason somehow specific to agoric-sdk, or would it apply and be useful to other users of @endo/exo? If so, 
- I'm sure you already know my feelings about introducing it this way to agoric-sdk. 
- But I understand not wanting to wait for another round of endo release and agoric-sdk sync. Attn @kriskowal @ivanlei 
- Please add a TODO at least here, that we expect this type to migrate to @endo/exo
- Please do an equivalent PR for @endo/exo, assuming that's the right place for `ExoObj` to eventually end up. Or at least file an issue with the need to do that and assign it to me.

What has me at a loss is that we cannot abstract away the location that `ExoObj` is imported from without loss of info in the IDE. Instead, each use location mentions `@agoric/internal`, all of which will need to change once we've completed the deprecation dance. I don't know what to suggest about that, but if you have any clever ideas, please let me know.

Reply:
I should also mention that since I don't yet understand why this pattern repeats or when it should be used, I don't yet have an opinion of whether `ExoObj` is a good name.

Reply:
@kriskowal when you say "not necessary to name" do you mean some other way to get `VaultDirector` from `prepareVaultDirector`? I'm having a hard time imagining how that's possible without a utility type. Codegen could, but that's unnecessarily complex.

Reply:
I withdraw my hope that we can avoid the need for a utility. For `ExoObj<T>`, what is `T` generally that requires this nested `Awaited` and `ReturnType` unboxing? Is this utility recursive or limited to a depth of two? Will there be a need for three levels? My hope is was that `T` was always an Exo (thus the name) and that we might change something about `T` to obviate the need for a utility.

Reply:
> What has me at a loss is that we cannot abstract away the location that ExoObj is imported from without loss of info in the IDE.

if you think that's better: 369ebece84213faf80cf8bf2381c16c0ab6fe769

> Instead, each use location mentions @agoric/internal, all of which will need to change once we've completed the deprecation dance

That doesn't seem so bad to me but I'm fine with putting it in the global namespace. If

Reply:
>For ExoObj<T>, what is T generally that requires this nested Awaited and ReturnType unboxing? 

The `prepare` function. 

>Is this utility recursive or limited to a depth of two?

Exactly two. `prepare` returns a `make` which returns an object. That's why I named it `ExoObj`. Open to other names.

With classes, the class definition specifies the constructor and classe's name is the name of the type for each instance. TS doesn't know about Exos so it doesn't make these names for us. I think this approach is the best we can hope for barring codegen.

Reply:
I see. `T` corresponds to an exo-class and you are inferring the exo-instance type. So, `Exo` at least should be in the name. May I suggest `ExoInstance<ExoClass>`? (Bearing in mind my opinions on Exos are barely informed and this might just get closer to right and not all the way there.)

Regardless, please document. I would hope this would be a welcome export from `@endo/exo`.

Reply:
I had `ExoInstance` at first but thought the shorter form would be worthwhile. The other two entities in the chain are both functions.

btw, it wouldn't be `<ExoClass>` because a class can be instantiated. The concrete type here is the `prepare`, which is a function to define a class (à la `defineExoClass`). Also this type works for `defineExoClassKit`. 

I haven't given thought to `makeExo` which makes a usable obj. I could make this type conditional on whether the param type is a callable function. 

> Regardless, please document

More than 57bce869a3e70879d82835bb386994b075a948e8 ?

Reply:
I looked, and the actual case is subtler than class vs instance. If it were class vs instance, we'd only need `ReturnType<ReturnsType<` without the `Awaited<`s. Rather, the direct class creation is wrapped with a function that defines one class per call to the function and returns some function making function, like the maker function returned by the class definition. If this wrapping function needs to `await` anything in its logic around the class creation, then the wrapping function is made async and it returns a promise for the instance making function.

If the instance making function is the one returned by the class definition, fine. But it itself might be a wrapper around the base instance making function. It still returns one instance each time it is called. But if any of the other things it does needs to `await` anything, then it returns a promise for the instance instead.

That accounts for both `Awaited<`s. When there are no promises in the way, the extra `Awaited<`s are harmless.

@turadg , is that right?

Reply:
> More than https://github.com/Agoric/agoric-sdk/commit/57bce869a3e70879d82835bb386994b075a948e8 ?

Yes please. There's a lot going on here!

Reply:
`makeExo` is at the same level as the second phase, of an instance making function that makes an instance each time it is called. For `makeExo` itself, we would only need `ReturnType<`. But the instance making function might be a wrapper around `makeExo` that does some other stuff around the call to `makeExo`. If any of that might need to be awaited, then the wrapping function will be async and we'd need the `Awaited<ReturnType<`. Does that make sense?

Reply:
>Does that make sense?

What you said in that paragraph matches my understanding, but I don't see how it supports the previous statement,
>Please don't fold makeExo into the same case. It isn't and should not look like it is.

I meant the result of a `makeExo` call, just as the others are results of `defineExo` calls. E.g., https://github.com/Agoric/agoric-sdk/blob/fd864a4d3fc254e640cbbdc7d8b1b158e197a356/packages/governance/src/question.js#L82-L100

So supporting,
```js
/** @typedef {ExoObj<typeof buildQuestion>`} Question
```

That has only one level of extracting the return type, vs two with the Exo class case.

Are you ok with the `ExoObj` supporting both levels? WDYT of Michael's idea to have the fully general `EReturn`?